SignalServiceKit/src/Messages/OWSMessageSender.m

```diff
@@ -954,14 +954,29 @@ - (void)sendMessage:(TSOutgoingMessage *)message
 
     NSString *localNumber = [TSAccountManager localNumber];
     if ([localNumber isEqualToString:recipient.uniqueId]) {
-        __block BOOL hasSecondaryDevices;
-        [self.dbConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
-            hasSecondaryDevices = [OWSDevice hasSecondaryDevicesWithTransaction:transaction];
-        }];
+        OWSAssert([message isKindOfClass:[OWSOutgoingSyncMessage class]]);
+        // Messages send to the "local number" should be sync messages.
+        //
+        // We can skip sending sync messages if we know that we have no linked
+        // devices. However, we need to be sure to handle the case where the
+        // linked device list has just changed.
+        //
+        // The linked device list is reflected in two separate pieces of state:
+        //
+        // * OWSDevice's state is updated when you link or unlink a device.
+        // * SignalRecipient's state is updated by 409 "Mismatched devices"
+        //   responses from the service.
+        //
+        // If _both_ of these pieces of state agree that there are no linked
+        // devices, then can safely skip sending sync message.
 
+        // 1. Check OWSDevice's state.
+        BOOL mayHaveLinkedDevices = [OWSDevice mayHaveLinkedDevices:self.dbConnection];
```

Comment:
There is still a concurrency concern here right? We're reading on a connection other than which we write with in the LinkDeviceController.

suggestion, which you can take or leave: move this method off the model class onto the DeviceService class, which is responsible for synchronizing access to `mayHaveLinkedDevice` (or a new tiny DeviceManager class).

Reply:
Okay, I've added caching which address this kind of race and will improve perf a tad.